File: sdk/android/rum/src/main/java/dev/openincident/rum/Event.kt
package dev.openincident.rum

import org.json.JSONObject

/**
 * One thing that happened in the application, in the shape the ingest reads.
 *
 * Built by hand into a `JSONObject` rather than serialised from the class: the
 * endpoint takes a small, stable, hand-written object, and generating it from
 * Kotlin property names means a rename in this file silently changes the
 * protocol. `org.json` is in the platform, so this costs no dependency.
 */
internal data class Event(
    val kind: Kind,
    val at: Long,
    val screen: String = "",
    val vital: String = "",
    val value: Double = 0.0,
    val rating: String = "",
    val errorType: String = "",
    val message: String = "",
    val stack: String = "",
    val attributes: Map<String, String> = emptyMap(),
) {
    internal enum class Kind(val wire: String) {
        SCREEN_VIEW("page_view"),
        VITAL("mobile_vital"),
        ERROR("error"),
        RESOURCE("resource"),
        ACTION("action"),
        LONG_TASK("long_task"),
    }

    fun payload(session: String, view: String, user: String?): JSONObject {
        val out = JSONObject()
        out.put("type", kind.wire)
        out.put("ts", at)
        out.put("session", session)
        out.put("view", view)
        // A screen has no URL. `route` is what the product groups by — one row
        // per screen rather than one per user — and a screen name is exactly
        // that, so it goes there and `url` stays empty.
        out.put("route", screen)
        out.put("url", "")
        if (vital.isNotEmpty()) {
            out.put("vital", vital)
            out.put("rating", rating)
        }
        // Outside a vital, `value` is how long a resource took, which is the
        // one thing worth knowing about a network call.
        if (value != 0.0) out.put("value", value)
        if (errorType.isNotEmpty()) out.put("errorType", errorType)
        if (message.isNotEmpty()) out.put("message", message)
        if (stack.isNotEmpty()) out.put("stack", stack)
        if (attributes.isNotEmpty()) out.put("attributes", JSONObject(attributes.toMap()))
        if (!user.isNullOrEmpty()) out.put("user", user)
        return out
    }
}

File: sdk/ios/Sources/OpenIncidentRUM/Event.swift
import Foundation

/// One thing that happened in the application, in the shape the ingest reads.
///
/// A struct with explicit JSON rather than `Codable` against the wire format:
/// the endpoint takes a small, stable, hand-written object, and generating it
/// from Swift property names means a rename in this file silently changes the
/// protocol.
struct Event {
    enum Kind: String {
        case screenView = "page_view"
        case vital = "mobile_vital"
        case error
        case resource
        case action
        case longTask = "long_task"
    }

    let kind: Kind
    let at: Date
    var screen: String = ""
    var vital: String = ""
    var value: Double = 0
    var rating: String = ""
    var errorType: String = ""
    var message: String = ""
    var stack: String = ""
    var attributes: [String: String] = [:]

    func payload(session: String, view: String, user: String?) -> [String: Any] {
        var out: [String: Any] = [
            "type": kind.rawValue,
            "ts": Int(at.timeIntervalSince1970 * 1000),
            "session": session,
            "view": view,
            // A screen has no URL. `route` is the field the product groups by —
            // one row per screen rather than one per visitor — and a screen
            // name is exactly that, so it goes there and `url` stays empty.
            "route": screen,
            "url": "",
        ]
        if !vital.isEmpty {
            out["vital"] = vital
            out["rating"] = rating
        }
        // Outside the vital, `value` is how long a resource took — the one
        // thing worth knowing about a network call. Sending it only alongside a
        // vital name dropped it silently, which is how the first run produced
        // rows saying a request happened and not that it took 143 ms.
        if value != 0 { out["value"] = value }
        if !errorType.isEmpty { out["errorType"] = errorType }
        if !message.isEmpty { out["message"] = message }
        if !stack.isEmpty { out["stack"] = stack }
        if !attributes.isEmpty { out["attributes"] = attributes }
        if let user, !user.isEmpty { out["user"] = user }
        return out
    }
}

/// The thresholds the two mobile timings are rated against, in milliseconds.
///
/// Carried in the event rather than applied when the chart is drawn, for the
/// same reason the browser SDK carries Google's: a rating that moved under a
/// stored row is a chart nobody can compare to last month.
enum Rating {
    static func of(_ vital: String, _ ms: Double) -> String {
        let bounds: (Double, Double)
        switch vital {
        case "app_start": bounds = (2000, 5000)
        case "screen_load": bounds = (1000, 2500)
        default: return ""
        }
        if ms <= bounds.0 { return "good" }
        if ms <= bounds.1 { return "needs-improvement" }
        return "poor"
    }
}
